benchmark/primitives/context_test.go

```diff
@@ -117,3 +117,34 @@ func BenchmarkTimerContextChannelGotErr(b *testing.B) {
 		}
 	}
 }
+
+type ctxKey struct{}
+
+func newContextWithLocalKey(parent context.Context) context.Context {
+	return context.WithValue(parent, ctxKey{}, nil)
+}
+
+var ck = ctxKey{}
+
+func newContextWithGlobalKey(parent context.Context) context.Context {
+	return context.WithValue(parent, ck, nil)
+}
+
+func BenchmarkContextWithValue(b *testing.B) {
+	benches := []struct {
+		name string
+		f    func(context.Context) context.Context
+	}{
+		{"newContextWithLocalKey", newContextWithLocalKey},
+		{"newContextWithGlobalKey", newContextWithGlobalKey},
+	}
+
+	pCtx := context.Background()
+	for _, bench := range benches {
+		b.Run(bench.name, func(b *testing.B) {
+			for j := 0; j < b.N; j++ {
```

Comment:
You may want to pass `b` and do the loop within the benchmark function.  Otherwise you are including the function call (if it isn't being inlined) with the cost, and the cost of this operation in general is low enough that it could be significant.

Reply:
I changed the code as follows:

```
type ctxKey struct{}

func newContextWithLocalKey(parent context.Context, b *testing.B) context.Context {
  var ret context.Context
  for i := 0; i < b.N; i++ {
    ret = context.WithValue(parent, ctxKey{}, nil)
  }
  return ret
}

var ck = ctxKey{}

func newContextWithGlobalKey(parent context.Context, b *testing.B) context.Context {
  var ret context.Context
  for i := 0; i < b.N; i++ {
    ret = context.WithValue(parent, ck, nil)
  }
  return ret
}

func BenchmarkContextWithValue(b *testing.B) {
  benches := []struct {
    name string
    f    func(context.Context, *testing.B) context.Context
  }{
    {"newContextWithLocalKey", newContextWithLocalKey},
    {"newContextWithGlobalKey", newContextWithGlobalKey},
  }

  pCtx := context.Background()
  rCtx := context.TODO()
  for _, bench := range benches {
    b.Run(bench.name, func(b *testing.B) { rCtx = bench.f(pCtx, b) })
  }

  if err := rCtx.Err(); err != nil {
    b.Errorf("got error from context {%v}, wanted an error-less context", err)
  }
}
```

The cost/op has increased for both, but still they are almost the same:: 

    easwars@minefield:grpc-go $ go test -bench=BenchmarkContextWithValue -test.benchmem google.golang.org/grpc/benchmark/primitives/...
    goos: linux
    goarch: amd64
    pkg: google.golang.org/grpc/benchmark/primitives
    BenchmarkContextWithValue/newContextWithLocalKey-12             20000000                83.4 ns/op            48 B/op          1 allocs/op
    BenchmarkContextWithValue/newContextWithGlobalKey-12            20000000                81.7 ns/op            48 B/op          1 allocs/op
    PASS
    ok      google.golang.org/grpc/benchmark/primitives     3.505s